subsys/sd/sdio.c

```diff
@@ -56,23 +63,995 @@ int sdio_send_ocr(struct sd_card *card, uint32_t ocr)
+			SDIO_CCCR_DRIVE_STRENGTH, 0, &data);
+		if (ret) {
+			return ret;
+		}
+		card->switch_caps.sd_drv_type = 0;
+		if (data & SDIO_CCCR_DRIVE_STRENGTH_A) {
+			card->switch_caps.sd_drv_type |= SD_DRIVER_TYPE_A;
+		}
+		if (data & SDIO_CCCR_DRIVE_STRENGTH_C) {
+			card->switch_caps.sd_drv_type |= SD_DRIVER_TYPE_C;
+		}
+		if (data & SDIO_CCCR_DRIVE_STRENGTH_D) {
+			card->switch_caps.sd_drv_type |= SD_DRIVER_TYPE_D;
+		}
+	}
+	return 0;
+}
+
+static void sdio_decode_cis(struct sdio_cis *cis, enum sdio_func_num func,
+			    uint8_t *data, uint8_t tpl_code, uint8_t tpl_link)
+{
+	switch (tpl_code) {
+	case SDIO_TPL_CODE_MANIFID:
+		cis->manf_id = data[0] | ((uint16_t)data[1] << 8);
+		cis->manf_code = data[2] | ((uint16_t)data[3] << 8);
+		break;
+	case SDIO_TPL_CODE_FUNCID:
+		cis->func_id = data[0];
+		break;
+	case SDIO_TPL_CODE_FUNCE:
+		if (func == 0) {
+			cis->max_blk_size = data[1] | ((uint16_t)data[2] << 8);
+			cis->max_speed = data[3];
+		} else {
+			cis->max_blk_size = data[12] | ((uint16_t)data[13] << 8);
+			cis->rdy_timeout = data[28] | ((uint16_t)data[29] << 8);
+		}
+		break;
+	default:
+		LOG_WRN("Unknown CIS tuple %d", tpl_code);
+		break;
+	}
+}
+
+/*
+ * Read CIS for a given SDIO function.
+ * Tuples provides a list of tuples that should be decoded.
+ */
+static int sdio_read_cis(struct sdio_func *func,
+			 uint8_t *tuples,
+			 uint32_t tuple_count)
+{
+	int ret;
+	char *data = func->card->card_buffer;
+	uint32_t cis_ptr, num = 0;
+	uint8_t tpl_code, tpl_link;
+	bool match_tpl = false;
+
+	memset(&func->cis, 0, sizeof(struct sdio_cis));
+	/* First find the CIS pointer for this function */
+	for (int i = 0; i < 3; i++) {
+		ret = sdio_io_rw_direct(func->card, SDIO_IO_READ, SDIO_FUNC_NUM_0,
+			SDIO_FBR_BASE(func->num) + SDIO_FBR_CIS + i, 0, data + i);
+		if (ret) {
+			return ret;
+		}
+	}
+	cis_ptr = *((uint32_t *)data);
+	/* Read CIS tuples until we have read all requested CIS tuple codes */
+	do {
+		/* Read tuple code */
+		ret = sdio_io_rw_direct(func->card, SDIO_IO_READ, func->num,
+			cis_ptr++, 0, &tpl_code);
+		if (ret) {
+			return ret;
+		}
+		if (tpl_code == SDIO_TPL_CODE_END) {
+			/* End of tuple chain */
+			break;
+		}
+		if (tpl_code == SDIO_TPL_CODE_NULL) {
+			/* Skip NULL tuple */
+			continue;
+		}
+		/* Read tuple link */
+		ret = sdio_io_rw_direct(func->card, SDIO_IO_READ, func->num,
+			cis_ptr++, 0, &tpl_link);
+		if (ret) {
+			return ret;
+		}
+		if (tpl_link == SDIO_TPL_CODE_END) {
+			/* End of tuple chain */
+			break;
+		}
+		/* Check to see if read tuple matches any we should look for */
+		for (int i = 0; i < tuple_count; i++) {
+			if (tpl_code == tuples[i]) {
+				match_tpl = true;
+				break;
+			}
+		}
+		if (match_tpl) {
+			/* tuple chains may be maximum of 255 bytes long */
+			memset(data, 0, 255);
+			for (int i = 0; i < tpl_link; i++) {
+				ret = sdio_io_rw_direct(func->card, SDIO_IO_READ,
+					func->num, cis_ptr++, 0, data + i);
```

Comment:
CIS info we must read by F0 (SDIO_FUNC_NUM_0):
`.. func->num,...`  -> `..SDIO_FUNC_NUM_0,...`